x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -331,12 +359,133 @@ public String toString() {
         return this.getClass().getSimpleName() + "@snapshotId=" + snapshotId + " lockFactory=" + lockFactory;
     }
 
+    private void prewarmCache() {
+        assert Thread.holdsLock(this);
+        assert loadCacheEagerly;
+        assert useCache;
+
+        final BlockingQueue<CheckedRunnable<Exception>> queue = new SizeBlockingQueue<>(new LinkedBlockingQueue<>(), Integer.MAX_VALUE);
+        for (BlobStoreIndexShardSnapshot.FileInfo file : snapshot().indexFiles()) {
+            final String fileName = file.physicalName();
+            if (isExcludedFromCache(fileName) == false && file.metadata().hashEqualsContents() == false) {
+                final long numberOfParts = file.numberOfParts();
+                if (queue.remainingCapacity() > numberOfParts) {
+                    logger.debug(
+                        "{} prewarming cache for file [{}] of length [{}] and [{}] parts",
+                        shardId,
+                        fileName,
+                        file.length(),
+                        numberOfParts
+                    );
+                    final long startTimeInNanos = statsCurrentTimeNanosSupplier.getAsLong();
+                    try {
+                        final IndexInput input = openInput(file, CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
```

Comment:
This method uses an `IndexInput` with a specific `IOContext` to prewarm the cache for the given Lucene file. The `IndexInput` will be cloned for each part to write in cache later and closed once all parts are processed.